core/services/relay/evm/mercury/v4/data_source.go

```diff
@@ -0,0 +1,290 @@
+}
+
+type LatestReportFetcher interface {
+	LatestPrice(ctx context.Context, feedID [32]byte) (*big.Int, error)
+	LatestTimestamp(context.Context) (int64, error)
+}
+
+type datasource struct {
+	pipelineRunner Runner
+	jb             job.Job
+	spec           pipeline.Spec
+	feedID         mercuryutils.FeedID
+	lggr           logger.Logger
+	saver          ocrcommon.Saver
+	orm            types.DataSourceORM
+	codec          reportcodec.ReportCodec
+
+	fetcher      LatestReportFetcher
+	linkFeedID   mercuryutils.FeedID
+	nativeFeedID mercuryutils.FeedID
+
+	mu sync.RWMutex
+
+	chEnhancedTelem chan<- ocrcommon.EnhancedTelemetryMercuryData
+}
+
+var _ v4.DataSource = &datasource{}
+
+func NewDataSource(orm types.DataSourceORM, pr pipeline.Runner, jb job.Job, spec pipeline.Spec, feedID mercuryutils.FeedID, lggr logger.Logger, s ocrcommon.Saver, enhancedTelemChan chan ocrcommon.EnhancedTelemetryMercuryData, fetcher LatestReportFetcher, linkFeedID, nativeFeedID mercuryutils.FeedID) *datasource {
+	return &datasource{pr, jb, spec, feedID, lggr, s, orm, reportcodec.ReportCodec{}, fetcher, linkFeedID, nativeFeedID, sync.RWMutex{}, enhancedTelemChan}
+}
+
+func (ds *datasource) Observe(ctx context.Context, repts ocrtypes.ReportTimestamp, fetchMaxFinalizedTimestamp bool) (obs v4types.Observation, pipelineExecutionErr error) {
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(ctx)
+
+	if fetchMaxFinalizedTimestamp {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			latest, dbErr := ds.orm.LatestReport(ctx, ds.feedID)
+			if dbErr != nil {
+				obs.MaxFinalizedTimestamp.Err = dbErr
+				return
+			}
+			if latest != nil {
+				maxFinalizedBlockNumber, decodeErr := ds.codec.ObservationTimestampFromReport(latest)
+				obs.MaxFinalizedTimestamp.Val, obs.MaxFinalizedTimestamp.Err = int64(maxFinalizedBlockNumber), decodeErr
+				return
+			}
+			obs.MaxFinalizedTimestamp.Val, obs.MaxFinalizedTimestamp.Err = ds.fetcher.LatestTimestamp(ctx)
+		}()
+	}
+
+	var trrs pipeline.TaskRunResults
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		var run *pipeline.Run
+		run, trrs, pipelineExecutionErr = ds.executeRun(ctx)
+		if pipelineExecutionErr != nil {
+			cancel()
+			pipelineExecutionErr = fmt.Errorf("Observe failed while executing run: %w", pipelineExecutionErr)
+			return
+		}
+
+		ds.saver.Save(run)
+
+		var parsed parseOutput
+		parsed, pipelineExecutionErr = ds.parse(trrs)
+		if pipelineExecutionErr != nil {
+			cancel()
+			// This is not expected under normal circumstances
+			ds.lggr.Errorw("Observe failed while parsing run results", "err", pipelineExecutionErr)
+			pipelineExecutionErr = fmt.Errorf("Observe failed while parsing run results: %w", pipelineExecutionErr)
+			return
+		}
+		obs.BenchmarkPrice = parsed.benchmarkPrice
+		obs.Bid = parsed.bid
+		obs.Ask = parsed.ask
+		obs.MarketStatus = parsed.marketStatus
```

Comment:
Where would we set MarketStatus.Err?

Reply:
The error is created by `setMarketStatus` and then returned by the `ds.parse` method.